### PR TITLE
feat: add shareable URL state for bookmarking and sharing

### LINKED ISSUE #108: Shareable URL for calculator state
### Context / Problem Statement

Users want to bookmark specific calculations or share them with colleagues. Currently, the app state is ephemeral - refreshing the page loses all configuration. There is no way to link to a specific model + image configuration.

### High-Level Description

Encode the current calculator state (selected model identifier + image configurations) into the URL hash so that calculations can be bookmarked, shared, and restored on page load. The URL updates as the user changes configuration, and loading a URL with state in the hash restores the calculator to that configuration.

### Requirements

1. The URL hash must encode: a schema version number, the selected model name (as a stable identifier), and for each image: height, width, multiplier, and preset name.
2. State must be serialized as compact JSON, then base64-encoded in the URL hash.
3. On page load, if the URL hash contains valid state, restore the model selection and image configuration, then run the calculation.
4. If a URL references a model name that no longer exists in the model list, show a warning and leave the model unselected.
5. Validate all restored values: dimensions must be non-negative finite numbers, multipliers must be >= 1, image count must be reasonable (cap at 50).
6. The URL must update when the user changes model selection or image configuration.
7. URL length must stay within safe browser limits (~2000 chars). If the encoded state exceeds this, show a warning that the URL may not be shareable and fall back to not updating the hash.
8. Include a "Copy link" button near the results that copies the current URL to clipboard.
9. Optionally include cost projection settings (requests per day) in the URL state if that feature is implemented.

### Expected Functionality

1. User configures model + images.
2. URL hash updates automatically (e.g. `#eyJ2IjoxLCJtIjoiR1BULTUuNCAoR2xvYmFsKSIs...`).
3. User copies URL and shares it.
4. Recipient opens URL - calculator restores with the same model, images, and results.
5. If the model in the URL no longer exists, a warning is shown.

### Constraints

- Model names are used as identifiers. If a model is renamed or removed, shared URLs referencing it will degrade gracefully (warning, not crash).
- URL state does not include uploaded image file data - only dimensions.

### Dependencies

- #104 - needed to cleanly run calculation on restored state.

### PR DESCRIPTION
**Pull request by jamesmcroft**
This pull request adds robust support for encoding and decoding calculator state in the URL, enabling users to share or restore their configuration via a link. It introduces a new `useUrlState` hook to synchronize state with the URL, provides user feedback for invalid or oversized URLs, and adds a "Copy Link" button in both single and comparison modes. Comprehensive unit tests are included to ensure correctness and handle edge cases.

**State Synchronization and URL Handling:**

* Introduced a new `useUrlState` hook to synchronize calculator state with the URL hash, restoring state on mount and updating the URL when state changes. The hook provides warnings for invalid/missing models and indicates when the configuration is too large to encode in the URL. (`src/hooks/useUrlState.js` [src/hooks/useUrlState.jsR1-R149](diffhunk://#diff-e2737fff7709c7de44a92f16735d99252665394127c8db9458334ad05a843578R1-R149))
* Implemented robust encoding and decoding logic for calculator state, including schema versioning, image validation, and support for comparison mode, with safeguards for oversized URLs and malformed data. (`src/utils/urlState.js` [src/utils/urlState.jsR1-R154](diffhunk://#diff-ed01c89690487aed80007b69222b9797846dd88ef030c018b3503296a4c59478R1-R154))
* Added comprehensive unit tests for URL state encoding, decoding, and model lookup, covering edge cases and validation logic. (`src/utils/__tests__/urlState.test.js` [src/utils/__tests__/urlState.test.jsR1-R251](diffhunk://#diff-6ce97544c4808d5d02d0a804bdcf33ba8b9a07475b5ff8b4fef91b7f2d52f471R1-R251))

**User Interface Enhancements:**

* Added user-facing alerts in the calculator UI to display warnings for invalid or missing models and to notify users when the configuration is too large to share via URL. (`src/components/calculator/Calculator.jsx` [[1]](diffhunk://#diff-cb2de9c5eccbe95ebc01b87c4597122a3a81253af0fc5503d09016e6575dcd4fR24-R25) [[2]](diffhunk://#diff-cb2de9c5eccbe95ebc01b87c4597122a3a81253af0fc5503d09016e6575dcd4fR86-R97)
* Added a "Copy Link" button to both single and comparison result views, allowing users to easily copy the current calculator URL. The button provides feedback on success or failure. (`src/components/calculator/CopyLinkButton.jsx` [[1]](diffhunk://#diff-6613a0f8e3e6ebf73a7390a7b038fd81ef70548ab83c17790f25d9aaa882b39aR1-R61) `src/components/calculator/CalculatorOutput.jsx` [[2]](diffhunk://#diff-25677943c5d70871ccc1e032a155f9a30d60f6068919bfd7b99887bf6bec8bceR84) `src/components/calculator/comparison/ComparisonTable.jsx` [[3]](diffhunk://#diff-3fe172f505ef7ee3092f68b34c08f41891dfec22096251eab2640884c3363b6fR52)

These changes significantly improve the shareability and resilience of the calculator by making it easy to share configurations and gracefully handling errors and edge cases.

### CODE REVIEW

File: src/hooks/useUrlState.js [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+    // Set state - single model (used when not in comparison mode)
+    if (modelObj && !result.comparisonMode) {
+      setModel(modelObj);
+      setModelName(modelObj.name);
+    }
+
+    // Clear existing images and add restored ones
+    const store = useBoundStore.getState();
+    while (store.images.length > 0) {
```
[the remark this answers is not in the document]
Comment (jamesmcroft) [reply src/hooks/useUrlState.js]: Fixed

File: src/hooks/useUrlState.js [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+  const setRequestsPerDay = useBoundStore((s) => s.setRequestsPerDay);
+  const comparisonMode = useBoundStore((s) => s.comparisonMode);
+  const setComparisonMode = useBoundStore((s) => s.setComparisonMode);
+  const selectedModels = useBoundStore((s) => s.selectedModels);
+  const toggleModelSelection = useBoundStore((s) => s.toggleModelSelection);
+
+  const hydratedRef = useRef(false);
+  const warningRef = useRef(null);
+  const oversizedRef = useRef(false);
+
```
[the remark this answers is not in the document]
Comment (jamesmcroft) [reply src/hooks/useUrlState.js:32]: Fixed

File: src/hooks/useUrlState.js [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+  const updateHash = useCallback(() => {
+    if (!hydratedRef.current) return;
+
+    const modelName = typeof model === "object" ? model?.name : "";
+    const hasContent = modelName || images.length > 0 || (comparisonMode && selectedModels.length > 0);
+    if (!hasContent) {
+      if (window.location.hash) {
+        history.replaceState(null, "", window.location.pathname);
+      }
```
[the remark this answers is not in the document]
Comment (jamesmcroft) [reply src/hooks/useUrlState.js:127]: Fixed

File: src/hooks/useUrlState.js [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+      useBoundStore.getState().removeImage(0);
+    }
+
+    for (const img of result.images) {
+      addImage(img);
+    }
+
+    if (result.requestsPerDay > 0) {
+      setRequestsPerDay(result.requestsPerDay);
+    }
```
[the remark this answers is not in the document]
Comment (jamesmcroft) [reply src/hooks/useUrlState.js]: Fixed

File: src/utils/__tests__/urlState.test.js [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+  it("omits preset when Custom", () => {
+    const { hash } = encodeState({
+      modelName: "Test",
+      images: [{ height: 100, width: 200, multiplier: 1, preset: "Custom" }],
+    }, testLocation);
+
+    // Decode the base64 to check the payload does not include p
+    const json = atob(hash.slice(1).replace(/-/g, "+").replace(/_/g, "/"));
+    const payload = JSON.parse(json);
+    expect(payload.i[0].p).toBeUndefined();
```
[the remark this answers is not in the document]
Comment (jamesmcroft) [reply src/utils/__tests__/urlState.test.js]: Fixed
